File: mobile/lib/core/shift_hours.dart
/// Studio shift helpers — mirrors backend IST rules (10:00–19:00).
const shiftStartHour = 10;
const shiftEndHour = 19;
const _istOffsetMinutes = 5 * 60 + 30;

String formatShiftHourLabel(int hour24) {
  final h = hour24 % 12 == 0 ? 12 : hour24 % 12;
  final suffix = hour24 >= 12 ? 'PM' : 'AM';
  return '$h $suffix';
}

String get shiftHoursLabel =>
    '${formatShiftHourLabel(shiftStartHour)} – ${formatShiftHourLabel(shiftEndHour)}';

/// Wall clock in studio timezone (Asia/Kolkata, UTC+5:30).
DateTime studioNow() {
  final utc = DateTime.now().toUtc();
  var totalMinutes = utc.hour * 60 + utc.minute + _istOffsetMinutes;
  var day = DateTime(utc.year, utc.month, utc.day);
  while (totalMinutes >= 24 * 60) {
    totalMinutes -= 24 * 60;
    day = day.add(const Duration(days: 1));
  }
  while (totalMinutes < 0) {
    totalMinutes += 24 * 60;
    day = day.subtract(const Duration(days: 1));
  }
  return DateTime(day.year, day.month, day.day, totalMinutes ~/ 60, totalMinutes % 60, utc.second);
}

DateTime studioShiftEndToday() {
  final now = studioNow();
  return DateTime(now.year, now.month, now.day, shiftEndHour);
}

Duration get expectedShiftDuration => Duration(hours: shiftEndHour - shiftStartHour);

DateTime studioShiftStartOnDay(DateTime day) {
  return DateTime(day.year, day.month, day.day, shiftStartHour);
}

DateTime studioShiftEndOnDay(DateTime day) {
  return DateTime(day.year, day.month, day.day, shiftEndHour);
}

Duration workedDuration({required DateTime clockIn, DateTime? clockOut}) {
  final end = clockOut ?? DateTime.now();
  final delta = end.difference(clockIn);
  return delta.isNegative ? Duration.zero : delta;
}

Duration lateStartDuration(DateTime clockIn) {
  final start = studioShiftStartOnDay(clockIn);
  if (!clockIn.isAfter(start)) return Duration.zero;
  return clockIn.difference(start);
}

Duration earlyEndDuration(DateTime clockOut) {
  final end = studioShiftEndOnDay(clockOut);
  if (!clockOut.isBefore(end)) return Duration.zero;
  return end.difference(clockOut);
}

/// Minutes still short of a full 9h shift (late start, early leave, or both).
Duration shiftOwedDuration({required DateTime clockIn, DateTime? clockOut}) {
  final worked = workedDuration(clockIn: clockIn, clockOut: clockOut);
  final shortfall = expectedShiftDuration - worked;
  if (shortfall.isNegative || shortfall == Duration.zero) return Duration.zero;
  return shortfall;
}

/// When a full shift must finish if they started late (clock-in + 9 hours).
DateTime fullShiftTargetTime(DateTime clockIn) => clockIn.add(expectedShiftDuration);

Duration timeUntilShiftEnd() {
  final now = studioNow();
  final end = studioShiftEndToday();
  if (!now.isBefore(end)) return Duration.zero;
  return end.difference(now);
}

Duration timeUntilFullShift(DateTime clockIn) {
  final target = fullShiftTargetTime(clockIn);
  final now = DateTime.now();
  if (!now.isBefore(target)) return Duration.zero;
  return target.difference(now);
}

String formatDurationHuman(Duration d) {
  if (d <= Duration.zero) return '0m';
  final h = d.inHours;
  final m = d.inMinutes.remainder(60);
  if (h > 0 && m > 0) return '${h}h ${m}m';
  if (h > 0) return '${h}h';
  return '${m}m';
}

String formatCountdown(Duration d) {
  final h = d.inHours;
  final m = d.inMinutes.remainder(60);
  final s = d.inSeconds.remainder(60);
  return '${h.toString().padLeft(2, '0')}:${m.toString().padLeft(2, '0')}:${s.toString().padLeft(2, '0')}';
}

String formatClockTime(DateTime dt) {
  final h = dt.hour % 12 == 0 ? 12 : dt.hour % 12;
  final m = dt.minute.toString().padLeft(2, '0');
  final suffix = dt.hour >= 12 ? 'PM' : 'AM';
  return '$h:$m $suffix';
}

File: frontend/src/lib/shiftHours.ts
/** Studio shift helpers — mirrors mobile + backend IST rules (10:00–19:00). */

export const SHIFT_START_HOUR = 10;
export const SHIFT_END_HOUR = 19;
export const STUDIO_TIMEZONE = "Asia/Kolkata";

export const shiftHoursLabel = `${formatShiftHourLabel(SHIFT_START_HOUR)} – ${formatShiftHourLabel(SHIFT_END_HOUR)}`;

export function formatShiftHourLabel(hour24: number) {
  const h = hour24 % 12 === 0 ? 12 : hour24 % 12;
  const suffix = hour24 >= 12 ? "PM" : "AM";
  return `${h} ${suffix}`;
}

type IstParts = {
  year: number;
  month: number;
  day: number;
  hour: number;
  minute: number;
  second: number;
};

export function istParts(date = new Date()): IstParts {
  const fmt = new Intl.DateTimeFormat("en-GB", {
    timeZone: STUDIO_TIMEZONE,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
    hour: "2-digit",
    minute: "2-digit",
    second: "2-digit",
    hour12: false,
  });
  const parts = Object.fromEntries(fmt.formatToParts(date).map((p) => [p.type, p.value]));
  const hourRaw = parts.hour === "24" ? "0" : parts.hour;
  return {
    year: Number(parts.year),
    month: Number(parts.month),
    day: Number(parts.day),
    hour: Number(hourRaw),
    minute: Number(parts.minute),
    second: Number(parts.second),
  };
}

export function studioNow(): Date {
  const p = istParts();
  return new Date(p.year, p.month - 1, p.day, p.hour, p.minute, p.second);
}

export function studioShiftEndToday(): Date {
  const now = studioNow();
  return new Date(now.getFullYear(), now.getMonth(), now.getDate(), SHIFT_END_HOUR, 0, 0, 0);
}

export const expectedShiftDurationMs = (SHIFT_END_HOUR - SHIFT_START_HOUR) * 60 * 60 * 1000;

export function studioShiftStartOnDay(day: Date): Date {
  return new Date(day.getFullYear(), day.getMonth(), day.getDate(), SHIFT_START_HOUR, 0, 0, 0);
}

export function studioShiftEndOnDay(day: Date): Date {
  return new Date(day.getFullYear(), day.getMonth(), day.getDate(), SHIFT_END_HOUR, 0, 0, 0);
}

export function workedDuration(clockIn: Date, clockOut?: Date | null): number {
  const end = clockOut ?? new Date();
  const delta = end.getTime() - clockIn.getTime();
  return delta < 0 ? 0 : delta;
}

export function lateStartDuration(clockIn: Date): number {
  const start = studioShiftStartOnDay(clockIn);
  if (clockIn.getTime() <= start.getTime()) return 0;
  return clockIn.getTime() - start.getTime();
}

export function earlyEndDuration(clockOut: Date): number {
  const end = studioShiftEndOnDay(clockOut);
  if (clockOut.getTime() >= end.getTime()) return 0;
  return end.getTime() - clockOut.getTime();
}

export function shiftOwedDuration(clockIn: Date, clockOut?: Date | null): number {
  const worked = workedDuration(clockIn, clockOut);
  const shortfall = expectedShiftDurationMs - worked;
  return shortfall <= 0 ? 0 : shortfall;
}

export function fullShiftTargetTime(clockIn: Date): Date {
  return new Date(clockIn.getTime() + expectedShiftDurationMs);
}

export function timeUntilShiftEnd(): number {
  const now = studioNow();
  const end = studioShiftEndToday();
  if (now.getTime() >= end.getTime()) return 0;
  return end.getTime() - now.getTime();
}

export function timeUntilFullShift(clockIn: Date): number {
  const target = fullShiftTargetTime(clockIn);
  const now = new Date();
  if (now.getTime() >= target.getTime()) return 0;
  return target.getTime() - now.getTime();
}

export function formatDurationHuman(ms: number): string {
  if (ms <= 0) return "0m";
  const totalMinutes = Math.floor(ms / 60_000);
  const h = Math.floor(totalMinutes / 60);
  const m = totalMinutes % 60;
  if (h > 0 && m > 0) return `${h}h ${m}m`;
  if (h > 0) return `${h}h`;
  return `${m}m`;
}

export function formatCountdown(ms: number): string {
  const totalSeconds = Math.max(0, Math.floor(ms / 1000));
  const h = Math.floor(totalSeconds / 3600);
  const m = Math.floor((totalSeconds % 3600) / 60);
  const s = totalSeconds % 60;
  return `${String(h).padStart(2, "0")}:${String(m).padStart(2, "0")}:${String(s).padStart(2, "0")}`;
}

export function formatClockTime(date: Date): string {
  return date.toLocaleTimeString(undefined, { hour: "numeric", minute: "2-digit", hour12: true });
}
